boot/seal.go

```diff
@@ -551,13 +559,46 @@ func resealFallbackObjectKeys(pbc predictableBootChains, authKeyFile string, rol
 	return nil
 }
 
+// recoveryModesForSystems returns a map for recovery modes for recovery systems
+// mentioned in the modeenv. The returned map contains both tested and candidate
+// recovery systems
+func modesForSystems(modeenv *Modeenv) map[string][]string {
+	if len(modeenv.GoodRecoverySystems) == 0 && len(modeenv.CurrentRecoverySystems) == 0 {
+		return nil
+	}
+
+	systemToModes := map[string][]string{}
+
+	// first go through tested recovery systems
+	modesForTestedSystem := []string{ModeRecover, ModeFactoryReset}
+	// tried systems can only boot to recovery mode
+	modesForCandidateSystem := []string{ModeRecover}
+
+	// go through recovery systems that were tested
+	for _, sys := range modeenv.GoodRecoverySystems {
+		systemToModes[sys] = modesForTestedSystem
+	}
+	// go through current recovery systems which can contain systems that
+	// are only being tried
+	for _, sys := range modeenv.CurrentRecoverySystems {
```

Comment:
can a system be in both Current and Good, in which case should we do first Current and then Good?

Reply:
Yes, a system is listed in both and that's a good point, I'll update the code.

Reply:
I see you changed the code, do we need a test for this case?